crates/sui-framework/src/natives/crypto/ecdsa_r1.rs

```diff
@@ -0,0 +1,134 @@
+};
+use smallvec::smallvec;
+use std::collections::VecDeque;
+use sui_types::error::SuiError;
+
+pub const FAIL_TO_RECOVER_PUBKEY: u64 = 0;
+pub const INVALID_SIGNATURE: u64 = 1;
+
+pub fn ecrecover(
+    _context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.is_empty());
+    debug_assert!(args.len() == 2);
+
+    let msg = pop_arg!(args, VectorRef);
+    let signature = pop_arg!(args, VectorRef);
+
+    let msg_ref = msg.as_bytes_ref();
+    let signature_ref = signature.as_bytes_ref();
+
+    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
+    let cost = legacy_empty_cost();
+    match recover_pubkey(&signature_ref, &msg_ref) {
+        Ok(pubkey) => Ok(NativeResult::ok(
+            cost,
+            smallvec![Value::vector_u8(pubkey.as_bytes().to_vec())],
+        )),
+        Err(SuiError::InvalidSignature { error: _ }) => {
+            Ok(NativeResult::err(cost, INVALID_SIGNATURE))
+        }
+        Err(_) => Ok(NativeResult::err(cost, FAIL_TO_RECOVER_PUBKEY)),
+    }
+}
+
+fn recover_pubkey(signature: &[u8], msg: &[u8]) -> Result<Secp256r1RecoverablePublicKey, SuiError> {
+    match <Secp256r1RecoverableSignature as ToFromBytes>::from_bytes(signature) {
+        Ok(signature) => match signature.recover(msg) {
+            Ok(pubkey) => Ok(pubkey),
+            Err(e) => Err(SuiError::KeyConversionError(e.to_string())),
+        },
+        Err(e) => Err(SuiError::InvalidSignature {
+            error: e.to_string(),
+        }),
+    }
+}
+
+pub fn secp256r1_verify(
+    _context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.is_empty());
+    debug_assert!(args.len() == 3);
+
+    let hashed_msg = pop_arg!(args, VectorRef);
+    let public_key_bytes = pop_arg!(args, VectorRef);
+    let signature_bytes = pop_arg!(args, VectorRef);
+
+    let hashed_msg_ref = hashed_msg.as_bytes_ref();
+    let public_key_bytes_ref = public_key_bytes.as_bytes_ref();
+    let signature_bytes_ref = signature_bytes.as_bytes_ref();
+
+    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/4086
+    let cost = legacy_empty_cost();
+
+    let signature = match <Secp256r1Signature as ToFromBytes>::from_bytes(&signature_bytes_ref) {
+        Ok(signature) => signature,
+        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
+    };
+
+    let public_key = match <Secp256r1PublicKey as ToFromBytes>::from_bytes(&public_key_bytes_ref) {
+        Ok(public_key) => public_key,
+        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
+    };
+
+    let result = public_key.verify(&hashed_msg_ref, &signature).is_ok();
+    Ok(NativeResult::ok(cost, smallvec![Value::bool(result)]))
+}
+
+pub fn secp256r1_verify_recoverable(
```

Comment:
what's the difference between this and the above function?

Reply:
the above verifies a nonrecoverable sig (64 bytes) vs this one verifies a recoverable one (65 bytes). they are documented at the move caller function signature instead of here since that's where the doc is generated.

Reply:
The verify function will recover the public key and compare it with the given public key, so if we want it this function, it could also be done using Move.